### PR TITLE
[Workload Management] QueryGroup Stats API Logic

### PR DESCRIPTION
**Pull request by ruai0511**
### Description
This change introduces the QueryGroup Stats API as part of this RFC: https://github.com/opensearch-project/OpenSearch/issues/12342. 

The QueryGroup Stats API schema is:
```
curl -XGET "localhost:9200/_wlm/stats"
```

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/action/admin/cluster/wlm/QueryGroupStatsResponse.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    QueryGroupStatsResponse(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    QueryGroupStatsResponse(ClusterName clusterName, List<QueryGroupStats> nodes, List<FailedNodeException> failures) {
+        super(clusterName, nodes, failures);
+    }
+
+    @Override
+    protected List<QueryGroupStats> readNodesFrom(StreamInput in) throws IOException {
```
Comment (kaushalmahi12) [inline server/src/main/java/org/opensearch/action/admin/cluster/wlm/QueryGroupStatsResponse.java:40]: Will this send the node_id along with the `QueryGroupStats`

File: server/src/main/java/org/opensearch/action/admin/cluster/wlm/TransportQueryGroupStatsAction.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        return queryGroupService.nodeStats();
+    }
+
+    /**
+     * Inner QueryGroupStatsRequest
+     *
+     * @opensearch.experimental
+     */
+    public static class NodeQueryGroupStatsRequest extends TransportRequest {
+
```
Comment (kaushalmahi12) [inline server/src/main/java/org/opensearch/action/admin/cluster/wlm/TransportQueryGroupStatsAction.java:92]: Do we really need this as this wrapper is not doing anything additional and the underlying object type is already `TransportRequest` ?

File: server/src/main/java/org/opensearch/client/ClusterAdminClient.java [hunk 3]
```diff
@@ -320,6 +322,14 @@ public interface ClusterAdminClient extends OpenSearchClient {
...
     NodesStatsRequestBuilder prepareNodesStats(String... nodesIds);
 
+    /**
+     * QueryGroup stats of the cluster.
+     *
+     * @param request  The QueryGroupStatsRequest
+     * @param listener A listener to be notified with a result
+     */
+    void queryGroupStats(QueryGroupStatsRequest request, ActionListener<QueryGroupStatsResponse> listener);
+
```
Comment (kaushalmahi12) [inline server/src/main/java/org/opensearch/client/ClusterAdminClient.java:332]: Why should we add this method in the ClusterAdmin interface ? I think this API should be accessible to all so does it make sense to add to `AbstractClient` since `NodeClient` extends it.

File: server/src/main/java/org/opensearch/rest/action/admin/cluster/RestQueryGroupStatsAction.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ * Transport action to get QueryGroup stats
+ *
+ * @opensearch.experimental
+ */
+public class RestQueryGroupStatsAction extends BaseRestHandler {
+
+    @Override
+    public List<Route> routes() {
+        return unmodifiableList(asList(new Route(GET, "/_wlm/query_group_stats"), new Route(GET, "/_wlm/query_group_stats/{nodeId}")));
+    }
```
Comment (kaushalmahi12) [inline server/src/main/java/org/opensearch/rest/action/admin/cluster/RestQueryGroupStatsAction.java:35]: Since this is core OS API we should remove the `_wlm` and keep it simple `query_group/stats` and it would be good if we could add queryQroupId as filter and other as mentioned in the stats issue: https://github.com/opensearch-project/OpenSearch/issues/15120

File: server/src/main/java/org/opensearch/wlm/QueryGroupService.java [hunk 5]
```diff
@@ -18,17 +20,22 @@
 
 /**
  * As of now this is a stub and main implementation PR will be raised soon.Coming PR will collate these changes with core QueryGroupService changes
+ * @opensearch.experimental
  */
 public class QueryGroupService {
     // This map does not need to be concurrent since we will process the cluster state change serially and update
     // this map with new additions and deletions of entries. QueryGroupState is thread safe
     private final Map<String, QueryGroupState> queryGroupStateMap;
+    private final TransportService transportService;
```
Comment (kaushalmahi12) [inline server/src/main/java/org/opensearch/wlm/QueryGroupService.java:29]: I think we should rely on just a `Supplier<String>` as we only need nodeId here

File: server/src/test/java/org/opensearch/action/support/nodes/TransportQueryGroupStatsActionTests.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        public TestTransportQueryGroupStatsAction(
+            ThreadPool threadPool,
+            ClusterService clusterService,
+            TransportService transportService,
+            QueryGroupService queryGroupService,
+            ActionFilters actionFilters
+        ) {
+            super(threadPool, clusterService, transportService, queryGroupService, actionFilters);
+        }
+    }
```
Comment (kaushalmahi12) [inline server/src/test/java/org/opensearch/action/support/nodes/TransportQueryGroupStatsActionTests.java:100]: I don't see a reason to have this class ? What is that this class do which original one doesn't

File: server/src/main/java/org/opensearch/client/ClusterAdminClient.java (hunk 3 above)
Comment (ruai0511) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:332]: the actual implementation is in `AbstractClient `. I see nodeStats is also following this pattern (abstract function in `ClusterAdminClient.java` and implementation in `AbstractClient`)

File: server/src/main/java/org/opensearch/rest/action/admin/cluster/RestQueryGroupStatsAction.java (hunk 4 above)
Comment (ruai0511) [reply server/src/main/java/org/opensearch/rest/action/admin/cluster/RestQueryGroupStatsAction.java:35]: Are you suggesting to turn `query_group/stats/{nodeId}` into `query_group/stats?nodeId=` and also support `query_group/stats?breach=`?

File: server/src/main/java/org/opensearch/action/admin/cluster/wlm/QueryGroupStatsResponse.java (hunk 1 above)
Comment (ruai0511) [reply server/src/main/java/org/opensearch/action/admin/cluster/wlm/QueryGroupStatsResponse.java:40]: This only returns List<QueryGroupStats> so no node_id

Comment (ruai0511) [reply server/src/main/java/org/opensearch/action/admin/cluster/wlm/QueryGroupStatsResponse.java:40]: This returns List<QueryGroupStats> but we can use queryGroupStats.getNode().getId() to get node_id

File: server/src/main/java/org/opensearch/action/admin/cluster/wlm/TransportQueryGroupStatsAction.java (hunk 2 above)
Comment (ruai0511) [reply server/src/main/java/org/opensearch/action/admin/cluster/wlm/TransportQueryGroupStatsAction.java:92]: I'm following the structure of `TransportNodesStatsAction`. `NodeQueryGroupStatsRequest` is used as a per-node action handler inside Transport Action, and QueryGroupStatsRequest represents the action for all nodes as a whole. The `NodeQueryGroupStatsRequest` serves as the container for the request sent to each node, allowing the framework to serialize and execute that request on the individual nodes.

**Comment by sgup432** [comment]
@ruai0511 @kaushalmahi12 
I guess the node stats URI should ```_wlm/query_group/stats```? As it looks more appropriate and also aligns with what we discussed in the stats meta issue. And allows for further extension in case we want.

**Comment by ruai0511** [comment]
> @ruai0511 @kaushalmahi12 I guess the node stats URI should `_wlm/query_group/stats`? As it looks more appropriate and also aligns with what we discussed in the stats meta issue. And allows for further extension in case we want.

If we use `GET _wlm/query_group/stats`, it'll create confusion between getting the stats and getting a querygroup named stats as the path would be the same

**Comment by sgup432** [comment]
>If we use GET _wlm/query_group/stats, it'll create confusion between getting the stats and getting a querygroup named stats as the path would be the same. So if we want to keep _wlm, we probably need to change it to _wlm/query_group_stats

I think it is upto user what kind of naming they want to use. In that case they can also use ```query_group_stats``` as a name which is weird and is also confusing in itself with current URI structure. 

I think having ```_wlm/<metric>/<name>/stats```  allows for better URI structure and is more consistent with the existing APIs.
If someone wants to use ```stats``` as a name so be it, they can still fetch it via ```_wlm/query_group/stats/stats``` which doesn't break the API from logical point of view.

**Review by kaushalmahi12 — APPROVED** [review]

File: server/src/main/java/org/opensearch/node/Node.java [hunk 7]
```diff
@@ -1026,29 +1026,6 @@ protected Node(
...
-            // register all standard SearchRequestOperationsCompositeListenerFactory to the SearchRequestOperationsCompositeListenerFactory
-            final SearchRequestOperationsCompositeListenerFactory searchRequestOperationsCompositeListenerFactory =
-                new SearchRequestOperationsCompositeListenerFactory(
-                    Stream.concat(
-                        Stream.of(
-                            searchRequestStats,
-                            searchRequestSlowLog,
-                            searchTaskRequestOperationsListener,
-                            queryGroupRequestOperationListener
-                        ),
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/node/Node.java:1045]: These changes look conflicting with Kaushal's PR

File: server/src/main/java/org/opensearch/wlm/QueryGroupService.java [hunk 8]
```diff
@@ -8,27 +8,39 @@
...
 public class QueryGroupService {
     // This map does not need to be concurrent since we will process the cluster state change serially and update
     // this map with new additions and deletions of entries. QueryGroupState is thread safe
     private final Map<String, QueryGroupState> queryGroupStateMap;
+    private final DiscoveryNode discoveryNode;
+    private final ClusterService clusterService;
 
-    public QueryGroupService() {
-        this(new HashMap<>());
+    @Inject
```
[the remark this answers is not in the document]
Comment (ruai0511) [reply server/src/main/java/org/opensearch/wlm/QueryGroupService.java:35]: Yeah we need this or else it'll throw error
```
 1) Could not find a suitable constructor in org.opensearch.wlm.QueryGroupService. Classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.
»    at org.opensearch.wlm.QueryGroupService.class(Unknown Source)
»    while locating org.opensearch.wlm.QueryGroupService
»      for parameter 3 at org.opensearch.action.admin.cluster.wlm.TransportQueryGroupStatsAction.<init>(Unknown Source)
»    at _unknown_

```

[the remark this answers is not in the document]
Comment (kaushalmahi12) [reply server/src/main/java/org/opensearch/wlm/QueryGroupService.java:35]: This basically means that you are using google Guice to instantiate the `TransportQueryGroupStatsAction` class. Maybe the ordering of bind operation is different but this is also fine.

File: server/src/main/java/org/opensearch/client/ClusterAdminClient.java [hunk 9]
```diff
@@ -320,6 +322,14 @@ public interface ClusterAdminClient extends OpenSearchClient {
      */
     NodesStatsRequestBuilder prepareNodesStats(String... nodesIds);
 
+    /**
+     * QueryGroup stats of the cluster.
+     *
+     * @param request  The QueryGroupStatsRequest
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/client/ClusterAdminClient.java:328]: nit: wlmStatsRequest

File: server/src/main/java/org/opensearch/node/Node.java [hunk 10]
```diff
@@ -1155,6 +1134,35 @@ protected Node(
...
+                                                                                                                                // this with
+                                                                                                                                // actual
+            // instance of the
+            // queryGroupService
+            queryGroupServiceSetOnce.set(queryGroupService);
+
+            final QueryGroupRequestOperationListener queryGroupRequestOperationListener = new QueryGroupRequestOperationListener(
+                queryGroupService,
+                threadPool
+            );
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/node/Node.java:1149]: I am assuming these changes are needed for collecting the stats? Ideally, the changes for collecting stats and the API should be in separate PRs. Also, @kaushalmahi12 is adding QueryGroupService as part of his PR. I am assuming one of you will need to resolve conflict depending on which PR is merged first.

File: server/src/main/java/org/opensearch/node/Node.java [hunk 11]
```diff
@@ -1155,6 +1134,35 @@ protected Node(
...
+                            searchRequestStats,
+                            searchRequestSlowLog,
+                            searchTaskRequestOperationsListener,
+                            queryGroupRequestOperationListener
+                        ),
+                        pluginComponents.stream()
+                            .filter(p -> p instanceof SearchRequestOperationsListener)
+                            .map(p -> (SearchRequestOperationsListener) p)
+                    ).toArray(SearchRequestOperationsListener[]::new)
+                );
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/node/Node.java:1165]: Why not initialize the `QueryGroupRequestOperationListener` and `QueryGroupService` before the original `SearchRequestOperationsCompositeListenerFactory` creation, instead of moving all the logic here?

File: server/src/main/java/org/opensearch/rest/action/admin/cluster/RestWlmStatsAction.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+                new Route(GET, "_wlm/{nodeId}/stats"),
+                new Route(GET, "_wlm/stats/{queryGroupId}"),
+                new Route(GET, "_wlm/{nodeId}/stats/{queryGroupId}")
+            )
+        );
+    }
+
+    @Override
+    public String getName() {
+        return "query_group_stats_action";
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/rest/action/admin/cluster/RestWlmStatsAction.java:47]: nit: wlm_stats_action

File: server/src/main/java/org/opensearch/wlm/WorkloadManagementTransportInterceptor.java [hunk 13]
```diff
@@ -20,11 +21,11 @@
...
 public class WorkloadManagementTransportInterceptor implements TransportInterceptor {
     private final ThreadPool threadPool;
-    private final QueryGroupService queryGroupService;
+    private final SetOnce<QueryGroupService> queryGroupService;
 
-    public WorkloadManagementTransportInterceptor(final ThreadPool threadPool, final QueryGroupService queryGroupService) {
+    public WorkloadManagementTransportInterceptor(final ThreadPool threadPool, final SetOnce<QueryGroupService> queryGroupServiceSetOnce) {
         this.threadPool = threadPool;
-        this.queryGroupService = queryGroupService;
+        this.queryGroupService = queryGroupServiceSetOnce;
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/wlm/WorkloadManagementTransportInterceptor.java:28]: Why do we need `SetOnce` here instead of usual class reference? I am assuming there is single `WorkloadManagementTransportInterceptor` instance per node

File: server/src/main/java/org/opensearch/wlm/stats/QueryGroupStats.java [hunk 14]
```diff
@@ -37,19 +39,26 @@
  *     ...
  * }
  */
-public class QueryGroupStats implements ToXContentObject, Writeable {
+public class QueryGroupStats extends BaseNodeResponse implements ToXContentObject, Writeable {
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/wlm/stats/QueryGroupStats.java:42]: I am wondering if we should rename this class also to `WlmStats` since, we might have more other stats in addition to `QueryGroupStats` within this class

File: server/src/main/java/org/opensearch/node/Node.java (hunk 11 above)
Comment (ruai0511) [reply server/src/main/java/org/opensearch/node/Node.java:1165]: The creation of `QueryGroupService` needs `transportService`, which is only initialized in line 1127. So we have to move `SearchRequestOperationsCompositeListenerFactory` down

**Comment by jainankitk** [comment]
@ruai0511 - Need to resolve conflicts from main branch after @kaushalmahi12's PR

**Comment by ruai0511** [comment]
> @ruai0511 - Need to resolve conflicts from main branch after @kaushalmahi12's PR

Just rebased. working on encapsulating querygroupstats now

**Comment by jainankitk** [comment]
@ruai0511 - Precommit is failing, can you address that?

File: server/src/main/java/org/opensearch/node/Node.java (hunk 11 above)
Comment (jainankitk) [reply server/src/main/java/org/opensearch/node/Node.java:1185]: We don't need `TransportService` into `QueryGroupService`. The `nodeStats` within `QueryGroupService` should be completely oblivious of the `localNodeId`. It should just return the `new QueryGroupStats(statsHolderMap)`. TransportAction can create the `WlmStats` object from `QueryGroupStats`.

File: server/src/main/java/org/opensearch/wlm/QueryGroupService.java [hunk 15]
```diff
@@ -205,16 +214,53 @@ public void incrementFailuresFor(final String queryGroupId) {
...
+                        .append(" limit is breaching for ENFORCED type QueryGroup: (")
+                        .append(threshold)
+                        .append(" < ")
+                        .append(lastRecordedUsage)
+                        .append("). ");
+                    return new Tuple<>(reason, resourceType);
+                }
+            }
+        }
+        return new Tuple<>(reason, null);
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/wlm/QueryGroupService.java:263]: Instead of reusing this method and create unnecessary reason objects in memory, we should write separate one that just returns the boolean for efficiency

**Review by jainankitk — CHANGES_REQUESTED** [review]

**Review by jainankitk — APPROVED** [review]

**Comment by jainankitk** [comment]
@ruai0511 - you need to merge from `main`:

```
org.opensearch.qa.verify_version_constants.VerifyVersionConstantsIT.testLuceneVersionConstant
Failing for the past 1 build (Since
[#49165](https://build.ci.opensearch.org/job/gradle-check/49165/) )
[Took 0.53 sec.](https://build.ci.opensearch.org/job/gradle-check/49165/testReport/junit/org.opensearch.qa.verify_version_constants/VerifyVersionConstantsIT/testLuceneVersionConstant/history)
Error Message

java.lang.AssertionError: 
Expected: <9.12.0>
     but: was <9.11.1>
```
